### PR TITLE
(feat) The lab order form should have a reason for ordering field

### PR DESCRIPTION
**Pull request by makombe**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Screenshots
lab-order-reason

## Related Issue

https://issues.openmrs.org/browse/O3-2361

## Other

Sample config schema

```ts
{
  "@openmrs/esm-patient-labs-app": {
    "labTestsWithOrderReasons": [
      {
        "labTestUuid": "854AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
        "orderReasons": [
          "161236AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
          "162080AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
          "160032AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
        ]
      }
    ]
  }
}
```

### CODE REVIEW

File: packages/esm-patient-labs-app/src/lab-orders/add-lab-order/lab-order-form.tsx
Comment (ibacher) [inline packages/esm-patient-labs-app/src/lab-orders/add-lab-order/lab-order-form.tsx:28]:
```suggestion
  const config = useConfig<ConfigObject>();
```

File: packages/esm-patient-labs-app/src/config-schema.ts
Comment (ibacher) [inline packages/esm-patient-labs-app/src/config-schema.ts:57]: Could we get a better description here? It's really unclear how this differs from `labTestUuid` just from the text and that description should be clear.

File: packages/esm-patient-labs-app/src/config-schema.ts
Comment (ibacher) [inline packages/esm-patient-labs-app/src/config-schema.ts:59]: As a general rule, we should avoid labels in configuration if we can. Labels like this are only usable in monolingual environments. In general, especially if this is concept-backed, we should prefer to use the display strings from the concepts themselves rather than inventing new text for them.

That said, what we store in the backend is just text, so I'm a little unclear on why we even have concepts involved in this at all? If we want to capture coded answers or coded or free text answers, we should actually update the backend and data model to reflect that.

Comment (ibacher) [reply packages/esm-patient-labs-app/src/config-schema.ts:59]: At a minimum, labels should be run through the translation system.

Comment (makombe) [reply packages/esm-patient-labs-app/src/config-schema.ts:59]: @ibacher orders model already provides the ability to capture order reason as coded or non coded/ free text. In this case we are using coded answers which are represented by concepts

Comment (ibacher) [reply packages/esm-patient-labs-app/src/config-schema.ts:59]: @makombe Thanks. I didn't realise it was using that pattern. If we're using concepts, then the label should come from the concept itself, right? Ideally, we want whatever we're using for drop-down labels to be translatable. Having strings in the configuration schema is, unfortunately, not translatable.

File: packages/esm-patient-labs-app/src/config-schema.ts
Comment (denniskigen) [inline packages/esm-patient-labs-app/src/config-schema.ts:49]:
```suggestion
        _description: "UUID for the lab test order that requires an order reason",
```

File: packages/esm-patient-labs-app/src/config-schema.ts
Comment (denniskigen) [inline packages/esm-patient-labs-app/src/config-schema.ts:57]:
```suggestion
            _description: 'Array of concept UUIDs that represent answers for an order reason concept',
```

File: packages/esm-patient-labs-app/src/config-schema.ts
Comment (denniskigen) [inline packages/esm-patient-labs-app/src/config-schema.ts:62]: Defer to Ian's line of reasoning above, but if we're to keep this property, a more appropriate description could be:

```suggestion
            _description: 'The label for order reason concept answer',
```

File: packages/esm-patient-labs-app/src/config-schema.ts
Comment (denniskigen) [inline packages/esm-patient-labs-app/src/config-schema.ts:70]: Not sure this accurately communicates what this config property does. It looks like we have an array that contains concepts for lab test orders that require a reason for ordering whose structure is something like:

```ts
// labTests array 
[ 
  {
    "labTestUuid": "52a447d3-a64a-11e3-9aeb-50e549534c5e",
    "labTestOrderReasons": [
      {
        "uuid": "52a447d3-a64a-11e3-9aeb-50e549534c5e",
        "label": "Test Reason 1"
      },
      {
        "uuid": "52a447d3-a64a-11e3-9aeb-50e549534c5e",
        "label": "Test Reason 2"
      }
    ] 
  }
]
```

Perhaps the config property would be better named `labTestOrdersWithReasonsForOrdering`.

File: packages/esm-patient-labs-app/src/config-schema.ts
Comment (denniskigen) [inline packages/esm-patient-labs-app/src/config-schema.ts:44]: Maybe?

```suggestion
  labTestsRequiringReasonsForOrdering: {
```

File: packages/esm-patient-labs-app/src/config-schema.ts
Comment (denniskigen) [inline packages/esm-patient-labs-app/src/config-schema.ts:52]: Could just be `reasonsForOrdering` because the array name already establishes the lab test order context.

```suggestion
      reasonsForOrdering: {
```

File: packages/esm-patient-labs-app/src/config-schema.ts
Comment (denniskigen) [inline packages/esm-patient-labs-app/src/config-schema.ts:66]:
```suggestion
        _description: 'Coded answers for order reason concepts',
```

**Comment by ojwanganto** [comment]
pinging @brandones

File: packages/esm-patient-labs-app/src/config-schema.ts
Comment (brandones) [reply packages/esm-patient-labs-app/src/config-schema.ts:59]: Agree with @ibacher here, we should be getting labels from the concepts, right?

Comment (brandones) [reply packages/esm-patient-labs-app/src/config-schema.ts:59]: Then we can simplify the config schema to be just an array of string concept UUIDs. Another alternative, which could be preferable (but I don't feel strongly about it, I'll leave it up to you) would be just to have `orderReasonsConceptSet`, which would be a UUID of a concept set that has the reasons as members.

File: packages/esm-patient-labs-app/src/lab-orders/add-lab-order/lab-order-form.component.tsx
Comment (brandones) [inline packages/esm-patient-labs-app/src/lab-orders/add-lab-order/lab-order-form.component.tsx:79]:
```suggestion
  const orderReasons = config.labTestsWithOrderReasons.find((c) => c.labTestUuid === defaultValues?.testType?.conceptUuid).map(c => c.orderReasons);
```

File: packages/esm-patient-labs-app/src/lab-orders/add-lab-order/lab-order-form.component.tsx
Comment (brandones) [inline packages/esm-patient-labs-app/src/lab-orders/add-lab-order/lab-order-form.component.tsx:197]:
```suggestion
          {orderReasons.length && (
```

File: packages/esm-patient-labs-app/src/config-schema.ts
Comment (ojwanganto) [reply packages/esm-patient-labs-app/src/config-schema.ts:59]: While this would be a good approach, it forces one to have concepts whose displays read exactly as the reasons provided in the drop-down. For instance, there could be a concept for indicating a baseline test with a display of **_initial_** while on the ground folks use **_baseline_** a lot. What would be the workaround for such? Do we support some degree of overrides?

Comment (brandones) [reply packages/esm-patient-labs-app/src/config-schema.ts:59]: Are you not able to change names on your concepts?

Comment (makombe) [reply packages/esm-patient-labs-app/src/config-schema.ts:59]: @brandones we normally try to avoid editing of concept directly as to be in sync with CIEL dictionary. We consume concepts as they come from CIEL.

File: packages/esm-patient-labs-app/src/lab-orders/add-lab-order/lab-order-form.component.tsx
Comment (makombe) [reply packages/esm-patient-labs-app/src/lab-orders/add-lab-order/lab-order-form.component.tsx]: updated.

File: packages/esm-patient-labs-app/src/config-schema.ts
Comment (ibacher) [reply packages/esm-patient-labs-app/src/config-schema.ts:59]: You can always request that CIEL adds a synonym or even preferred name for a concept. Alternatively, you can use Iniz to provide custom concept names or change the preferred name in a reproducible way without needing to deviate from CIEL.

In any case, labels _must never_ appear in the configuration because it is unworkable in a multilingual environment. If we need something in the configuration to represent a user-facing string it needs to be a translation key rather than a label.

Comment (ojwanganto) [reply packages/esm-patient-labs-app/src/config-schema.ts:59]: Great points, thanks. I know adding synonyms in CIEL may be good but still not sustainable more so if preferences/synonyms differ across implementations. A major challenge we face with labels is trying to match the labels as they are provided in clinical forms or even reporting tools. Even though the concept is the same, we are forced to display it differently depending on the type of user, and how they commonly use the concept. 

The concept to use the iniz module is great but I don't see how it will address the label overrides for concepts that may require 2 or 3 synonyms. I may be wrong, but I was hoping that the idea behind a config is to somehow address a local context. For instance, I don't expect the same config to apply in Kenya and Tanzania where English and Kiswahili are used. I would expect the Kenyanized config to be in English and that for Tanzania in Kiswahili. Sorry if I may have missed the point.

Comment (makombe) [reply packages/esm-patient-labs-app/src/config-schema.ts]: @ibacher  @brandones I have removed the label from the config as suggested and made necessary adjustment. Thanks

Comment (ojwanganto) [reply packages/esm-patient-labs-app/src/config-schema.ts]: Thanks @makombe. @ibacher @brandones @denniskigen kindly help with review. Thanks

**Comment by ojwanganto** [comment]
@denniskigen kindly help us review this. This PR will unblock the use of o3 lab order form for HIV monitoring lab tests

**Review by denniskigen — APPROVED** [review]

**Comment by brandones** [comment]
Thank you @denniskigen for finishing the review, thanks @ojwanganto and @makombe for your work!

**Pull request merged**
